File: app/src/main/java/com/vidyutgati/core/battery/VoltageSagFilter.kt
package com.vidyutgati.core.battery

/**
 * Exponential Moving Average (EMA) filter to dampen sudden voltage sag
 * during high throttle acceleration or climbing gradients.
 *
 * Prevents false "Battery Critical" alerts when an e-rickshaw pulls a heavy 4-passenger load from a stop.
 */
class VoltageSagFilter(
    private val smoothingFactor: Double = 0.2 // Weight given to new sample (0.0 to 1.0)
) {
    private var smoothedVoltage: Double? = null

    /**
     * Filters an incoming raw voltage sample and returns the stable smoothed voltage.
     */
    fun filter(rawVoltage: Double): Double {
        val current = smoothedVoltage
        return if (current == null) {
            smoothedVoltage = rawVoltage
            rawVoltage
        } else {
            val updated = (smoothingFactor * rawVoltage) + ((1.0 - smoothingFactor) * current)
            smoothedVoltage = updated
            updated
        }
    }

    /**
     * Resets the filter, useful when switching battery chemistry or starting a new shift.
     */
    fun reset(initialVoltage: Double? = null) {
        smoothedVoltage = initialVoltage
    }

    fun getCurrentSmoothedVoltage(): Double? = smoothedVoltage
}

File: iosApp/VidyutGati/Core/VoltageSagFilter.swift
import Foundation

/**
 * Exponential Moving Average (EMA) filter to dampen sudden voltage sag
 * during high throttle acceleration or climbing gradients on iOS.
 */
public final class IOSVoltageSagFilter {
    private let smoothingFactor: Double
    private var smoothedVoltage: Double?

    public init(smoothingFactor: Double = 0.25) {
        self.smoothingFactor = smoothingFactor
    }

    public func filter(rawVoltage: Double) -> Double {
        if let current = smoothedVoltage {
            let updated = (smoothingFactor * rawVoltage) + ((1.0 - smoothingFactor) * current)
            smoothedVoltage = updated
            return updated
        } else {
            smoothedVoltage = rawVoltage
            return rawVoltage
        }
    }

    public func reset(initialVoltage: Double? = nil) {
        smoothedVoltage = initialVoltage
    }

    public func getCurrentSmoothedVoltage() -> Double? {
        return smoothedVoltage
    }
}
